evap/evaluation/models.py

```diff
@@ -1,34 +1,332 @@
+import threading
 import uuid
-import operator
 
 from django.conf import settings
 from django.contrib import messages
 from django.contrib.auth.models import AbstractBaseUser, BaseUserManager, Group, PermissionsMixin
+from django.contrib.contenttypes.fields import GenericForeignKey
+from django.contrib.contenttypes.models import ContentType
 from django.contrib.postgres.fields import ArrayField
 from django.core.cache import caches
-from django.core.exceptions import ValidationError
+from django.core.exceptions import FieldDoesNotExist, ValidationError
 from django.core.mail import EmailMessage
-from django.db import models, transaction, IntegrityError
-from django.db.models import Count, Q, Manager
+from django.db import IntegrityError, models, transaction
+from django.db.models import Count, Manager, Q
+from django.db.models.signals import m2m_changed
 from django.dispatch import Signal, receiver
+from django.forms.models import model_to_dict
 from django.template import Context, Template
 from django.template.base import TemplateSyntaxError
+from django.template.loader import render_to_string
+from django.urls import reverse
 from django.utils import timezone
+from django.utils.formats import localize
 from django.utils.functional import cached_property
 from django.utils.translation import gettext_lazy as _
-from django.urls import reverse
 from django_fsm import FSMField, transition
 from django_fsm.signals import post_transition
-
-from evap.evaluation.tools import clean_email, date_to_datetime, translate, is_external_email
+from evap.evaluation.tools import clean_email, date_to_datetime, is_external_email, translate
 
 logger = logging.getLogger(__name__)
 
+FieldAction = namedtuple("FieldAction", "label type items")
+
+
+class LogEntry(models.Model):
+    content_type = models.ForeignKey(ContentType, on_delete=models.CASCADE, related_name="logs_about_me")
+    content_object_id = models.PositiveIntegerField(db_index=True)
+    content_object = GenericForeignKey("content_type", "content_object_id")
+    attached_to_object_type = models.ForeignKey(ContentType, on_delete=models.CASCADE, related_name="logs_for_me")
+    attached_to_object_id = models.PositiveIntegerField(db_index=True)
+    attached_to_object = GenericForeignKey("attached_to_object_type", "attached_to_object_id")
+    datetime = models.DateTimeField(auto_now_add=True)
+    user = models.ForeignKey(settings.AUTH_USER_MODEL, null=True, blank=True, on_delete=models.PROTECT)
+    action_type = models.CharField(max_length=255)
+    request_id = models.CharField(max_length=36, null=True, blank=True)
+    data = models.TextField(default="{}")
+
+    class Meta:
+        ordering = ("-datetime", "-id")
+
+    @staticmethod
+    def _pk_to_string_representation(key, field, related_objects):
+        if key is None:
+            return key
+        try:
+            return str(related_objects.get(pk=key))
+        except field.related_model.DoesNotExist:
+            return "�"
```

Comment:
If the m2m set changes on an object, we store the primary keys of the objects added/deleted. For display, we fetch those objects and build a string representation from them. We can't do that if those objects got deleted, but we might still want to show that *an* object as deleted. I figured this symbol could represent an "unknown" object. 

With evap, I think this will rarely be seen, as we rarely delete objects used on m2m fields with Evaluation and Course (that's mostly Questionaires and users), but I still included this as a sensible last resort.